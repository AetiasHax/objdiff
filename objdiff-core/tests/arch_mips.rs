use objdiff_core::{diff, obj};

mod common;

#[test]
#[cfg(feature = "mips")]
fn read_mips() {
    let diff_config = diff::DiffObjConfig { mips_register_prefix: true, ..Default::default() };
    let obj = obj::read::parse(include_object!("data/mips/main.c.o"), &diff_config).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "ControlEntry").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}
