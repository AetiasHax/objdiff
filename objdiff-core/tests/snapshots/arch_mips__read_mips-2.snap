---
source: objdiff-core/tests/arch_mips.rs
expression: diff.instruction_rows
---
[
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 0,
                size: 4,
                opcode: 12,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 4,
                size: 4,
                opcode: 44,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 8,
                size: 4,
                opcode: 44,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 12,
                size: 4,
                opcode: 44,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 16,
                size: 4,
                opcode: 44,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 20,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 24,
                size: 4,
                opcode: 113,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 28,
                size: 4,
                opcode: 26,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 32,
                size: 4,
                opcode: 20,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 36,
                size: 4,
                opcode: 97,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 40,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 44,
                size: 4,
                opcode: 12,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 48,
                size: 4,
                opcode: 20,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 52,
                size: 4,
                opcode: 26,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 56,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 60,
                size: 4,
                opcode: 12,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 64,
                size: 4,
                opcode: 26,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 68,
                size: 4,
                opcode: 97,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 72,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 76,
                size: 4,
                opcode: 97,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 80,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: Some(
            InstructionBranchFrom {
                ins_idx: [
                    22,
                ],
                branch_idx: 0,
            },
        ),
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 84,
                size: 4,
                opcode: 97,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 88,
                size: 4,
                opcode: 56,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 20,
                branch_idx: 0,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 92,
                size: 4,
                opcode: 113,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 96,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 100,
                size: 4,
                opcode: 20,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 104,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 108,
                size: 4,
                opcode: 12,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 112,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 116,
                size: 4,
                opcode: 16,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 120,
                size: 4,
                opcode: 20,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 124,
                size: 4,
                opcode: 12,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 128,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: Some(
            InstructionBranchFrom {
                ins_idx: [
                    36,
                    38,
                    44,
                ],
                branch_idx: 1,
            },
        ),
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 132,
                size: 4,
                opcode: 12,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 136,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 140,
                size: 4,
                opcode: 113,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 144,
                size: 4,
                opcode: 55,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 32,
                branch_idx: 1,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 148,
                size: 4,
                opcode: 90,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 152,
                size: 4,
                opcode: 3,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 32,
                branch_idx: 1,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 156,
                size: 4,
                opcode: 113,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 160,
                size: 4,
                opcode: 2,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 164,
                size: 4,
                opcode: 60,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 168,
                size: 4,
                opcode: 77,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 172,
                size: 4,
                opcode: 113,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 176,
                size: 4,
                opcode: 54,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 32,
                branch_idx: 1,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 180,
                size: 4,
                opcode: 113,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
]
